controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -320,51 +323,66 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
         long maxExecutionPeriod = Math.min(MAX_EXECUTION_TIME_MULTIPLIER * lease, Duration.ofDays(1).toMillis());
 
-        UUID txnId = UUID.randomUUID();
-        TxnResource resource = new TxnResource(scope, stream, txnId);
-
-        // Step 2. Add txn to host-transaction index.
-        CompletableFuture<Void> addIndex = validate.thenComposeAsync(ignore ->
-                streamMetadataStore.addTxnToIndex(hostId, resource, 0), executor).whenComplete((v, e) -> {
-                    if (e != null) {
-                        log.debug("Txn={}, failed adding txn to host-txn index of host={}", txnId, hostId);
-                    } else {
-                        log.debug("Txn={}, added txn to host-txn index of host={}", txnId, hostId);
-                    }
-                });
-
-        // Step 3. Create txn node in the store.
-        CompletableFuture<VersionedTransactionData> txnFuture = addIndex.thenComposeAsync(ignore ->
-                streamMetadataStore.createTransaction(scope, stream, txnId, lease, maxExecutionPeriod,
-                        scaleGracePeriod, ctx, executor), executor).whenComplete((v, e) -> {
-                    if (e != null) {
-                        log.debug("Txn={}, failed creating txn in store", txnId);
-                    } else {
-                        log.debug("Txn={}, created in store", txnId);
-                    }
-                });
+        // 1. get latest epoch from history
+        // 2. generateNewTransactionId.. this step can throw WriteConflictException
+        // 3. txn id = 32 bit epoch + 96 bit counter
+        // 4. if while creating txn epoch no longer exists, then we will get DataNotFoundException.
+        // 5. Retry if we get WriteConflict or DataNotFoundException, from step 1.
+        // Note: this is a low probability for either exceptions:
+        // - WriteConflict, if multiple controllers are trying to get new range at the same time then we can get write conflict
+        // - DataNotFoundException because it will only happen in rare case
+        // when we generate the transactionid against latest epoch (if there is ongoing scale then this is new epoch)
+        // and then epoch node is deleted as scale starts and completes.
+        return validate.thenCompose(validated -> RetryHelper.withRetriesAsync(() ->
+                streamMetadataStore.generateTransactionId(scope, stream, ctx, executor)
+                .thenCompose(txnId -> {
+                    TxnResource resource = new TxnResource(scope, stream, txnId);
+                    // Step 2. Add txn to host-transaction index.
+                    CompletableFuture<Void> addIndex = streamMetadataStore.addTxnToIndex(hostId, resource, 0)
+                            .whenComplete((v, e) -> {
+                                if (e != null) {
+                                    log.debug("Txn={}, failed adding txn to host-txn index of host={}", txnId, hostId);
```

Comment:
Don't you want to rethrow the exception? Otherwise your downstream task will execute regardless

Reply:
whencomplete does not swallow up the exception.

The documentation of whenComplete(https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#whenComplete-java.util.function.BiConsumer-) says:

```    Returns a new CompletionStage with the same result or exception as this stage, that executes the given action when this stage completes.```